File: demos/barbershop/Android/app/src/main/java/dev/mobiler/barbershop/TransferPlugin.kt
package dev.mobiler.barbershop

import dev.mobiler.barbershop.shared.types.HttpHeader
import dev.mobiler.barbershop.shared.types.HttpOutcome
import dev.mobiler.barbershop.shared.types.PluginResponse
import dev.mobiler.barbershop.shared.types.TransferEvent

import android.app.Application
import android.net.Uri
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import okio.Buffer
import okio.BufferedSink
import okio.buffer
import okio.sink
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.io.InputStream

/**
 * Streaming file upload/download (free, bundled). See `mobiler-plugin.toml` + the iOS twin
 * (`ios/TransferPlugin.swift`) for the manifest. Rides the streaming primitive:
 *   cx.upload(url, source).start("up", Msg::Xfer)   // PUT (default)/POST source → url
 *   cx.download(url, dest).start("dl", Msg::Xfer)    // GET url → dest
 *   cx.unsubscribe(key)                              // cancels
 * Envelope JSON (see mobiler-core::transfer::TransferReq): {"url","source"|"dest","method"?,
 * "headers":[{"name","value"}]}. `op` is "upload" or "download" (the plugin op, not the HTTP verb).
 *
 * CANCEL SEMANTICS (must match iOS — see TransferPlugin.swift): `cx.unsubscribe` cancels the
 * collecting coroutine. Whether that lands before the transfer starts or mid-flight, we emit
 * NOTHING for it — the app already unsubscribed and isn't listening for a terminal event (this
 * mirrors the web shell's silent pre-send cancel; see mobiler-web's `TransferHandle::drop`). A
 * `cancelled` flag (set in `awaitClose`, which runs before `call.cancel()` reaches the network
 * layer) distinguishes that deliberate teardown from a genuine transport failure, which DOES emit
 * `Done{TransportError}`.
 */
class TransferPlugin(private val application: Application) : MobilerPlugin {
    private val client = OkHttpClient()

    // Transfer only makes sense as a stream (progress + terminal event) — `handle` exists only to
    // satisfy the MobilerPlugin interface, same as the built-in ticker/system streams.
    override suspend fun handle(op: String, input: String): PluginResponse =
        PluginResponse(false, "transfer is a streaming capability — use cx.subscribe")

    override fun subscribe(op: String, input: String): Flow<PluginResponse> = callbackFlow {
        val obj = runCatching { JSONObject(input) }.getOrNull()
        if (obj == null) {
            trySend(eventResponse(TransferEvent.Done(HttpOutcome.TransportError("malformed transfer envelope"), null)))
            close()
            return@callbackFlow
        }
        val url = obj.optString("url")
        val builder = Request.Builder().url(url)
        // Content-Type is threaded through as `appContentType` instead of added as a plain header
        // here: OkHttp's BridgeInterceptor writes the upload RequestBody's `contentType()` as the
        // Content-Type header, REPLACING anything set via `addHeader` — so an app-supplied value
        // added here would be silently overwritten by streamingUploadBody's fixed
        // "application/octet-stream". Filtering it out and letting `contentType()` be the single
        // source of truth (app value if present, else octet-stream) keeps this in sync with iOS
        // (`request.addValue`) and web (`xhr.set_request_header`), which both honor it.
        var appContentType: String? = null
        obj.optJSONArray("headers")?.let { hs ->
            for (i in 0 until hs.length()) {
                val h = hs.getJSONObject(i)
                val name = h.optString("name")
                if (op == "upload" && name.equals("Content-Type", ignoreCase = true)) {
                    appContentType = h.optString("value")
                } else {
                    builder.addHeader(name, h.optString("value"))
                }
            }
        }

        // Set by `awaitClose` (see below) the instant unsubscribe fires, BEFORE `call.cancel()`
        // reaches OkHttp — so any completion/failure callback that races with it always observes
        // `cancelled == true` first and stays silent.
        var cancelled = false
        var call: Call? = null
        var lastEmit = 0L

        // ~10/sec progress throttle, purely on elapsed time (same rule as the web shell). The
        // terminal Done is never throttled — it's sent unconditionally by its own call site.
        fun maybeProgress(transferred: Long, total: Long?) {
            val now = System.currentTimeMillis()
            if (now - lastEmit < 100) return
            lastEmit = now
            trySend(eventResponse(TransferEvent.Progress(transferred.toULong(), total?.toULong())))
        }

        fun finish(ev: TransferEvent.Done) {
            if (!cancelled) trySend(eventResponse(ev))
            close()
        }

        if (op == "upload") {
            val method = obj.optString("method").ifEmpty { "PUT" }
            val source = obj.optString("source")
            val body = streamingUploadBody(application, source, appContentType, ::maybeProgress)
            if (body == null) {
                finish(TransferEvent.Done(HttpOutcome.TransportError("cannot open upload source '$source'"), null))
                return@callbackFlow
            }
            val request = builder.method(method, body).build()
            val c = client.newCall(request)
            call = c
            c.enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    finish(TransferEvent.Done(HttpOutcome.TransportError(e.message ?: "upload failed"), null))
                }
                override fun onResponse(call: Call, response: Response) {
                    response.use { r ->
                        val headers = r.headers.map { (n, v) -> HttpHeader(n, v) }
                        finish(TransferEvent.Done(HttpOutcome.Response(r.code.toUShort(), headers, emptyList()), null))
                    }
                }
            })
        } else {
            val destPath = obj.optString("dest")
            val dest = resolveSandboxPath(application, destPath)
            if (dest == null) {
                finish(TransferEvent.Done(HttpOutcome.TransportError("bad dest '$destPath'"), null))
                return@callbackFlow
            }
            val request = builder.build() // no explicit method → OkHttp defaults to GET
            val c = client.newCall(request)
            call = c
            // Launched as a CHILD coroutine (not run inline here) so this producer body can reach
            // `awaitClose` immediately and register the cancellation hook concurrently with the
            // blocking read below — required for `call.cancel()` to actually interrupt a stalled
            // `source.read()` (a plain blocking Okio call that does not itself check cancellation).
            launch(Dispatchers.IO) {
                try {
                    c.execute().use { response ->
                        val respBody = response.body
                        val total = respBody?.contentLength()?.takeIf { it >= 0 }
                        dest.parentFile?.mkdirs()
                        var got = 0L
                        if (respBody != null) {
                            respBody.source().use { src ->
                                dest.sink().buffer().use { sink ->
                                    val buf = Buffer()
                                    while (true) {
                                        // Cooperative cancellation for the common case (steady
                                        // stream, frequent small reads); `call.cancel()` (fired
                                        // from `awaitClose` on unsubscribe) is the fallback that
                                        // unblocks a `read()` stalled waiting on the socket.
                                        ensureActive()
                                        val n = src.read(buf, 64L * 1024)
                                        if (n == -1L) break
                                        sink.write(buf, n)
                                        got += n
                                        maybeProgress(got, total)
                                    }
                                }
                            }
                        }
                        val headers = response.headers.map { (n, v) -> HttpHeader(n, v) }
                        finish(TransferEvent.Done(HttpOutcome.Response(response.code.toUShort(), headers, emptyList()), Uri.fromFile(dest).toString()))
                    }
                } catch (e: CancellationException) {
                    dest.delete()
                    throw e // never swallow — let structured concurrency see the cancellation
                } catch (e: IOException) {
                    dest.delete()
                    finish(TransferEvent.Done(HttpOutcome.TransportError(e.message ?: "download failed"), null))
                }
            }
        }

        awaitClose {
            cancelled = true
            call?.cancel()
        }
    }
}

/**
 * Builds a streaming, re-openable upload `RequestBody` for `source`: a `content://` URI (via the
 * ContentResolver — length usually unknown) or a sandbox-relative/absolute/`file://` path (length
 * known upfront). Returns null if the source cannot be opened at all (probed once, eagerly, so a
 * bad path fails fast instead of only inside `writeTo`). `onProgress` is called after every chunk
 * write with (bytesSentSoFar, totalOrNull).
 */
private fun streamingUploadBody(
    application: Application,
    source: String,
    contentType: String?,
    onProgress: (sent: Long, total: Long?) -> Unit,
): RequestBody? {
    val isContentUri = source.startsWith("content://")
    val sourceFile: File? = if (isContentUri) null else resolveSandboxPath(application, source)
    if (!isContentUri && sourceFile == null) return null

    fun open(): InputStream? =
        if (isContentUri) application.contentResolver.openInputStream(Uri.parse(source))
        else sourceFile!!.takeIf { it.exists() }?.inputStream()

    // Eager probe: fail fast on a missing/unreadable source rather than only inside writeTo
    // (which OkHttp calls later, off this thread, where a null-body error is harder to report).
    open()?.close() ?: return null

    val total: Long? = if (isContentUri) null else sourceFile!!.length().takeIf { it > 0 }
    // App-supplied Content-Type wins (needed e.g. for a presigned S3/GCS PUT signed against a
    // specific content type); otherwise default to octet-stream. This is the single place the
    // upload's Content-Type header is decided — see the caller in `subscribe` above.
    val mediaType: MediaType? = (contentType ?: "application/octet-stream").toMediaTypeOrNull()

    return object : RequestBody() {
        override fun contentType(): MediaType? = mediaType
        override fun contentLength(): Long = total ?: -1L
        override fun writeTo(sink: BufferedSink) {
            val input = open() ?: throw IOException("cannot open source '$source'")
            input.use { ins ->
                var sent = 0L
                val chunk = ByteArray(64 * 1024)
                while (true) {
                    val n = ins.read(chunk)
                    if (n == -1) break
                    sink.write(chunk, 0, n)
                    sent += n
                    onProgress(sent, total)
                }
            }
        }
    }
}

/** Resolve a sandbox-relative path (or an absolute/`file://` path) to a File; reject `..` escapes
 *  outside the app's private filesDir. Mirrors `files` plugin's `resolve`. */
private fun resolveSandboxPath(application: Application, path: String): File? {
    if (path.startsWith("file://")) return Uri.parse(path).path?.let { File(it) }
    if (path.startsWith("/")) return File(path)
    val root = application.filesDir
    val f = File(root, path)
    return if (f.canonicalPath.startsWith(root.canonicalPath)) f else null
}

/** ok = true for a Progress tick, or a Done whose outcome is a 2xx Response (matches HttpPlugin's
 *  `ok` = 2xx convention; a non-2xx response and a TransportError are both `ok = false`). */
private fun eventResponse(ev: TransferEvent): PluginResponse {
    val ok = when (ev) {
        is TransferEvent.Progress -> true
        is TransferEvent.Done -> {
            val outcome = ev.outcome
            outcome is HttpOutcome.Response && outcome.status.toInt() in 200..299
        }
    }
    return PluginResponse(ok, ev.bincodeSerialize().toUByteList())
}

/** The generated types use List<UByte>, not List<Byte> — ByteArray.toList() gives the wrong
 *  element type and will not compile. */
private fun ByteArray.toUByteList(): List<UByte> = this.map { it.toUByte() }

File: mobiler/plugins/transfer/ios/TransferPlugin.swift
import Foundation
import SharedTypes

/// Streaming file upload/download (free, bundled). See `mobiler-plugin.toml` + the Android twin
/// (`android/TransferPlugin.kt`) for the manifest. Rides the streaming primitive:
///   cx.upload(url, source).start("up", Msg.xfer)   // PUT (default)/POST source → url
///   cx.download(url, dest).start("dl", Msg.xfer)    // GET url → dest
///   cx.unsubscribe(key)                             // cancels
/// Envelope JSON (see mobiler-core::transfer::TransferReq): {"url","source"|"dest","method"?,
/// "headers":[{"name","value"}]}. `op` is "upload" or "download" (the plugin op, not the HTTP verb).
///
/// CANCEL SEMANTICS (must match Android — see TransferPlugin.kt): `cx.unsubscribe` cancels this
/// subscription's Task. Whether that lands before the URLSessionTask is even created, before it's
/// resumed, or mid-flight, we emit NOTHING for it — the app already unsubscribed and isn't
/// listening for a terminal event (this mirrors the web shell's silent pre-send cancel; see
/// mobiler-web's `TransferHandle::drop`). `TransferDelegate.cancelledByApp`, flipped by the
/// `onCancel` handler before `task.cancel()` runs, distinguishes that deliberate teardown from a
/// genuine transport failure, which DOES emit `Done{TransportError}`.
enum TransferPlugin {
    static func handle(op: String, input: String) async -> PluginResponse {
        PluginResponse(ok: false, output: "transfer is a streaming capability — use cx.subscribe")
    }

    static func subscribe(op: String, input: String, emit: @escaping @Sendable (PluginResponse) -> Void) async {
        guard
            let data = input.data(using: .utf8),
            let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let urlString = obj["url"] as? String,
            let url = URL(string: urlString)
        else {
            emit(response(for: .done(outcome: .transportError(message: "malformed transfer envelope"), handle: nil)))
            return
        }

        var request = URLRequest(url: url)
        if let headers = obj["headers"] as? [[String: Any]] {
            for h in headers {
                guard let name = h["name"] as? String, let value = h["value"] as? String else { continue }
                request.addValue(value, forHTTPHeaderField: name)
            }
        }

        // Resolves a sandbox-relative path, an absolute path, or a `file://` URI to a URL. iOS has
        // no `content://` scheme (that's Android-only), so this is simpler than the Android twin.
        func resolve(_ path: String) -> URL {
            path.hasPrefix("file://") ? (URL(string: path) ?? URL(fileURLWithPath: path)) : URL(fileURLWithPath: path)
        }

        let uploadSource: URL?
        let downloadDest: URL?
        if op == "upload" {
            request.httpMethod = (obj["method"] as? String) ?? "PUT"
            guard let source = obj["source"] as? String else {
                emit(response(for: .done(outcome: .transportError(message: "missing upload source"), handle: nil)))
                return
            }
            uploadSource = resolve(source)
            downloadDest = nil
        } else {
            guard let dest = obj["dest"] as? String else {
                emit(response(for: .done(outcome: .transportError(message: "missing download dest"), handle: nil)))
                return
            }
            uploadSource = nil
            downloadDest = resolve(dest)
        }

        let delegate = TransferDelegate(emit: emit, dest: downloadDest)
        // A dedicated session (not .shared) so this transfer's delegate/lifetime is self-contained,
        // matching the websocket plugin's `URLSession(configuration: .default)` precedent.
        let session = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
        let task: URLSessionTask
        if op == "upload", let uploadSource {
            task = session.uploadTask(with: request, fromFile: uploadSource)
        } else {
            task = session.downloadTask(with: request)
        }

        await withTaskCancellationHandler {
            await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
                // Cancel race (mirrors mobiler-web's `start_web_upload` comment): `onCancel` below
                // can fire before this closure runs at all (the enclosing Task was already
                // cancelled the instant it was created). Checking here, right before `resume()`,
                // is what keeps that case silent — the task never starts, so no delegate callback
                // ever fires and nothing is ever emitted for it.
                if Task.isCancelled {
                    delegate.markCancelledByApp()
                    cont.resume()
                    return
                }
                delegate.onFinish = { cont.resume() }
                task.resume()
            }
        } onCancel: {
            delegate.markCancelledByApp()
            task.cancel()
        }
        session.finishTasksAndInvalidate()
    }

    fileprivate static func response(for ev: TransferEvent) -> PluginResponse {
        let ok: Bool
        switch ev {
        case .progress: ok = true
        case .done(let outcome, _):
            if case .response(let status, _, _) = outcome { ok = (200..<300).contains(Int(status)) } else { ok = false }
        }
        return PluginResponse(ok: ok, output: (try? ev.bincodeSerialize()) ?? [])
    }
}

/// Bridges URLSessionTaskDelegate/URLSessionDownloadDelegate callbacks to `emit`. One instance per
/// transfer (unlike the websocket plugin's single long-lived connection actor) — a transfer is a
/// one-shot request, not a persistent connection. `@unchecked Sendable`: all mutable state here is
/// only ever touched from URLSession's delegate queue (serial by default) plus the two call sites
/// in `subscribe` above that run strictly before/after the session is alive, so there's no
/// concurrent access — same reasoning the bluetooth plugin's `BleManager` uses.
private final class TransferDelegate: NSObject, URLSessionTaskDelegate, URLSessionDownloadDelegate, @unchecked Sendable {
    private let emit: @Sendable (PluginResponse) -> Void
    private let dest: URL? // set only for downloads — where to move the finished temp file
    private var lastEmit = Date.distantPast
    private var cancelledByApp = false
    private var finished = false
    /// Resumes the continuation `subscribe` is suspended on. Set right before `task.resume()`.
    var onFinish: (() -> Void)?

    init(emit: @escaping @Sendable (PluginResponse) -> Void, dest: URL?) {
        self.emit = emit
        self.dest = dest
    }

    func markCancelledByApp() { cancelledByApp = true }

    // ~10/sec progress throttle (same rule as the web shell / Android twin). The terminal Done is
    // never throttled — each completion delegate method below emits it unconditionally.
    private func maybeEmitProgress(transferred: Int64, total: Int64) {
        let now = Date()
        guard now.timeIntervalSince(lastEmit) >= 0.1 else { return }
        lastEmit = now
        let t: UInt64? = total > 0 ? UInt64(total) : nil
        emit(TransferPlugin.response(for: .progress(transferred: UInt64(max(0, transferred)), total: t)))
    }

    // Upload progress.
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
        maybeEmitProgress(transferred: totalBytesSent, total: totalBytesExpectedToSend)
    }

    // Download progress.
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        maybeEmitProgress(transferred: totalBytesWritten, total: totalBytesExpectedToWrite)
    }

    // Download completion: move the system's temp file to `dest` (overwriting), then emit Done
    // with the response headers/status. `didCompleteWithError` also fires right after this (with
    // `error == nil`) — the `finished` guard there skips it so we don't double-emit.
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        finished = true
        defer { onFinish?() }
        guard let dest else { return } // shouldn't happen — a download always sets `dest`
        do {
            try? FileManager.default.removeItem(at: dest) // overwrite
            try FileManager.default.createDirectory(at: dest.deletingLastPathComponent(), withIntermediateDirectories: true)
            try FileManager.default.moveItem(at: location, to: dest)
        } catch {
            try? FileManager.default.removeItem(at: dest)
            if !cancelledByApp {
                emit(TransferPlugin.response(for: .done(outcome: .transportError(message: "failed to save: \(error.localizedDescription)"), handle: nil)))
            }
            return
        }
        guard !cancelledByApp else { return }
        let http = downloadTask.response as? HTTPURLResponse
        let status = UInt16(clamping: http?.statusCode ?? 0)
        let headers = (http?.allHeaderFields ?? [:]).compactMap { key, value -> HttpHeader? in
            guard let name = key as? String else { return nil }
            return HttpHeader(name: name, value: String(describing: value))
        }
        emit(TransferPlugin.response(for: .done(outcome: .response(status: status, headers: headers, body: []), handle: dest.absoluteString)))
    }

    // Fires for both uploads and downloads. For downloads this is a NO-OP when the transfer
    // succeeded (`didFinishDownloadingTo` already emitted Done and set `finished`); it's still the
    // only place a download's cancel/failure surfaces (the temp file is cleaned up by the OS, so
    // there's no partial `dest` to delete on a failed/cancelled download — `dest` was never
    // touched). For uploads this is the ONLY completion path.
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard !finished else { return }
        finished = true
        defer { onFinish?() }
        guard !cancelledByApp else { return }
        if let error {
            emit(TransferPlugin.response(for: .done(outcome: .transportError(message: error.localizedDescription), handle: nil)))
            return
        }
        // Upload success.
        let http = task.response as? HTTPURLResponse
        let status = UInt16(clamping: http?.statusCode ?? 0)
        let headers = (http?.allHeaderFields ?? [:]).compactMap { key, value -> HttpHeader? in
            guard let name = key as? String else { return nil }
            return HttpHeader(name: name, value: String(describing: value))
        }
        emit(TransferPlugin.response(for: .done(outcome: .response(status: status, headers: headers, body: []), handle: nil)))
    }
}
